### DISCUSSION TITLE
Custom conversions for CursorResult

### DISCUSSION THREAD

**Discussion by iljau** [sqlalchemy/sqlalchemy#8545] [Usage Questions]
In current version of Snowflake driver, fields with following types: "ARRAY", "VARIANT", "OBJECT" are returned in raw form and need to be passed though json.loads to be converted to Python types.

Metadata for that is available by looking at cursor description.
```
conn: sqlalchemy.engine.base.Connection
result = conn.execute(sql)
cdesc = result.cursor.description
 ```
 
That can be done in separate function which would return list of dicts with conversions applied.

What I'd like to have instead, is having this custom logic run so that returned CursorResult would already have these conversions applied. What is unclear is how this can be accomplished (e.g. what to subclass, where to override some methods).

**Comment by zzzeek** [comment]
are you modifying the Snowflake dialect directly or are you looking for an end-user workaround?  The way dialects do this is by using dialect specific type objects that have result processor functions set up.   on the end-user side you can build these functions using TypeDecorator: https://docs.sqlalchemy.org/en/14/core/custom_types.html#augmenting-existing-types   for snowflake, they'd want to add result_processor() methods to the types at https://github.com/snowflakedb/snowflake-sqlalchemy/blob/main/src/snowflake/sqlalchemy/custom_types.py

**Comment by iljau** [comment]
"end-user workaround" is what I'm looking for (using `snowflake-sqlalchemy` and/or `snowflake-connector-python` as is, no influence over these);

Looking at TypeDecorator docs it's not immediately clear how to use it (and how it applies to non-orm `conn.execute` code), but I found this example: https://docs.sqlalchemy.org/en/14/core/custom_types.html#marshal-json-strings - which if works would solve the problem. Will have to try it out.

**Reply by zzzeek** [reply]
> "end-user workaround" is what I'm looking for (using `snowflake-sqlalchemy` and/or `snowflake-connector-python` as is, no influence over these);

if those drivers are doing the wrong thing in some case, you should report a bug, everyone has influence over this kind of project.


> 
> Looking at TypeDecorator docs it's not immediately clear how to use it (and how it applies to non-orm `conn.execute` code), but I found this example: https://docs.sqlalchemy.org/en/14/core/custom_types.html#marshal-json-strings - which if works would solve the problem. Will have to try it out.

the type objects apply if you are querying with select() or orm.Query() objects.  they dont normally apply if you are running a SQL string directly unless you use the text() construct.
